commit 10d7d8a33c4f19b238fa2f49aaa41f122d072fd5
Author: Ben Gamari <ben@smart-cactus.org>
Date:   Tue Oct 9 14:41:08 2018 -0400

    Introduce suppport for NIX_DEBUG_INFO_DIRS environment var
    
    Look up .build-id files relative to the directories in the
    colon-separated environment variable NIX_DEBUG_INFO_DIRS, in addition
    to the existing debug-file-directory setting.

diff --git a/gdb/build-id.c b/gdb/build-id.c
index c8eacbd1e8..9bb5140e3b 100644
--- a/gdb/build-id.c
+++ b/gdb/build-id.c
@@ -67,8 +67,8 @@ build_id_verify (bfd *abfd, size_t check_len, const bfd_byte *check)
 
 /* See build-id.h.  */
 
-gdb_bfd_ref_ptr
-build_id_to_debug_bfd (size_t build_id_len, const bfd_byte *build_id)
+static gdb_bfd_ref_ptr
+build_id_to_debug_bfd_in (const char *directories, size_t build_id_len, const bfd_byte *build_id)
 {
   gdb_bfd_ref_ptr abfd;
 
@@ -76,7 +76,7 @@ build_id_to_debug_bfd (size_t build_id_len, const bfd_byte *build_id)
      cause "/.build-id/..." lookups.  */
 
   std::vector<gdb::unique_xmalloc_ptr<char>> debugdir_vec
-    = dirnames_to_char_ptr_vec (debug_file_directory);
+    = dirnames_to_char_ptr_vec (directories);
 
   for (const gdb::unique_xmalloc_ptr<char> &debugdir : debugdir_vec)
     {
@@ -123,6 +123,30 @@ build_id_to_debug_bfd (size_t build_id_len, const bfd_byte *build_id)
   return abfd;
 }
 
+gdb_bfd_ref_ptr
+build_id_to_debug_bfd (size_t build_id_len, const bfd_byte *build_id)
+{
+  gdb_bfd_ref_ptr abfd = build_id_to_debug_bfd_in(debug_file_directory, build_id_len, build_id);
+
+  if (abfd != NULL)
+    return abfd;
+
+  static int init = 0;
+  static char *env_var;
+  if (!init)
+    {
+      env_var = getenv("NIX_DEBUG_INFO_DIRS");
+      init = 1;
+    }
+
+  if (env_var)
+    {
+      abfd = build_id_to_debug_bfd_in(env_var, build_id_len, build_id);
+    }
+
+  return abfd;
+}
+
 /* See build-id.h.  */
 
 std::string
diff --git a/gdb/symfile.c b/gdb/symfile.c
index 6a1140edbc..538d429ede 100644
--- a/gdb/symfile.c
+++ b/gdb/symfile.c
@@ -1351,8 +1351,8 @@ show_debug_file_directory (struct ui_file *file, int from_tty,
 			   struct cmd_list_element *c, const char *value)
 {
   fprintf_filtered (file,
-		    _("The directory where separate debug "
-		      "symbols are searched for is \"%s\".\n"),
+		    _("The directories where separate debug "
+		      "symbols are searched for are \"%s\".\n"),
 		    value);
 }
 
